### PR TITLE
[WIP] feat: harden post type usage and establish relationships between listings and posts/pages

### LINKED ISSUE #33: Harden post type usage
We built Newspack Listings to be flexible in the usage of the various listing types, but in order to more easily cross-reference listings with each other and other content types, we should more clearly define the use cases for Places and Marketplace listings: businesses should always be represented as places, and Marketplace listings as one-off pieces of content. This would make it easier for us to create a parent-child relationship between businesses as Places and the Marketplace listings they own.

If we did this, Places would become the de facto content type that publishers sell to third parties to represent their business, and Marketplace listings would be the content that those third parties could create and edit on their own. Marketplace listings could also serve as classified ads and obituaries, and we would offer block pattern layouts for all of these use cases, while removing the Business patterns from the Marketplace post type.

We could also create a similar parent-child relationship between Places and Events, so that (for example) a concert venue could be created as a Place, and Event listings that take place at the venue could be associated with it. Porting ideas from #22 here (not vetted, just ideas):

* If a Place is added to an Event as a venue, when viewing the Event's single page, you should see some info on the venue appear automatically below the content.
* If that Event is added to a list with map functionality enabled, its venue's location data should be added to the map (as well as any map data existing in the Event post content itself).
* If viewing a Place that is attached to one or more Events, you should be able to see a list of the Events below the content.

### LINKED ISSUE #34: Direct associations with Posts and Pages
Create a way to build relationships between all the different listing types and posts/pages. One way to do this would be to create a metadata field in the sidebar for posts and pages that lets editors search for listings to associate with that content. This could be very similar to the UI we show for searching for listings in a Specific Posts curated list block, but for all listing types, not just one at a time:

listing-search-ui

After one or more listings is associated with a post or page, we would have the ability to e.g. show them on the front-end in a custom related content module.

### PR DESCRIPTION
**Pull request by dkoo**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This one is kind of big! Does a few things to lay the groundwork for establishing cross-data-type relationships:

1. Restricts the use of Business Listing patterns to the Place listing post type. This establishes Places as the de facto business data type, and Marketplace listings as more one-off classifieds which can be "owned" by one or more Places.
2. Creates shadow taxonomies for each listing post type. These taxonomies are automatically created, updated, or deleted when posts of the corresponding listing types are created, renamed, or unpublished, respectively.
3. Shadow taxonomies can be assigned to various post types. This lets us create associations between posts of different types—for example, Places can be assigned to Marketplace listings and Events. Listings can also be associated with other listings of the same type, which should let us show "related listings" more easily. This closes #33. 
4. Each shadow taxonomy can be assigned to specific listings of other types, and all can be assigned to pages and posts. Closes #34.
5. When viewing a post, page, or listing that's been assigned listing shadow terms, we can automatically show a blurb and link for those associated listings. I've implemented a test version of this using the existing `.author-bio` styles from the Newspack theme, but this might need more design consideration (cc @thomasguillot).
6. Adds a couple of test block patterns for Marketplace listings (in the form of real estate listings). These are definitely placeholder patterns that I created mainly for testing purposes, so we'll want to revisit in another PR (cc @thomasguillot).
7. Adds a custom "Price" block that I figured would be useful for Marketplace listings (e.g. classifieds and real estate listings). This block editor is a fairly simple number input that also lets you select from a list of WooCommerce currencies (if available—defaults to the site's default currency, and falls back to USD if WooCommerce is not available) and then formats the number shown on the front-end with the corresponding currency symbol and internationalization based on the editor's browser settings—so if you choose Pound Sterling and you as an admin have a browser set to `en-GB`, it should show the number 10000 as `£10,000.00`. If you choose Euro and your browser is set to `fr-FR`, it should show `10 000,00 EUR`. I based the formatting on the admin's browser settings because I figured that the price should be shown in the format familiar to the person creating the listing and choosing the currency. This block is also synced to a single meta field for listing CPTs, which automatically inherits the value from first instance of a Price block. This should be useful for future search and filtering features.

#### Testing shadow taxonomies

1. After checking out this branch, edit a post or page—you should now see sidebar panels for each listing post type alongside categories and tags. Also confirm that terms are automatically created for pre-existing published listings (so editors who are already running listings won't need to go back and update or recreate those to get the new terms).

Screen Shot 2021-03-31 at 4 31 51 PM

2. Edit listings of each type and confirm that you see sidebar panels for the listing post types designated as being assignable in `class-newspack-listings-taxonomies.php`. For example, you should be able to assign Events to other Events and Generic listings, Generic listings only to other Generic listings, Marketplace listings to Generic listings and other Marketplace listings, and Places to all listing types.
3. Assign some listings to posts, pages, and other listings and view the front-end. 
4. You should see blurbs for each assigned listing at the bottom of the content (see below):

**Note:** I'm not sure the language ("Listed by", "Related listings") is quite right, but I'm also not sure what these headings should say. It seems particularly weird to show "Listed by" when some of the blurbs are of the same type (e.g., showing an Event blurb on another Event), but I'm also not sure what we should show in the case of related listings of the same type. The blurbs also visually clash with author bios and sponsor blurbs, if those are enabled. Open to suggestions.

For example, on a post or page:
Screen Shot 2021-03-31 at 4 31 34 PM

On another listing:
Screen Shot 2021-03-31 at 4 25 36 PM

#### Testing shadow taxonomy updating/deleting

1. Publish a new listing of any type.
2. Edit a post, page, or another listing that is able to receive that listing type as a shadow term.
3. Confirm that the new listing is searchable/selectable as a term in the editor sidebar.
4. Revert the listing to a draft.
5. Refresh the editor for the post/page/other listing and confirm that the listing term is no longer selectable or searchable.
6. Publish a new post of any type other than a listing and confirm that no new term is created under any of the listing shadow taxonomies.

#### Testing Price block

1. Create or edit a Marketplace listing.
2. Confirm that a new block called "Price" is now insertable through the Block Inserter. Conversely, confirm that the Price block is not available in any other post type.

Screen Shot 2021-03-31 at 6 48 48 PM

3. Test the block functionality:
  - Enter numbers in the input field. Confirm that you can only enter numbers greater than or equal to 0.
  - Confirm that the Currency field defaults to the currency set for your site in WooCommerce settings (which is set during Newspack onboarding). If you want to go above and beyond, you can install Listings on a fresh site without WooCommerce and confirm that a.) the Currency field is hidden, and b.) formatted prices are only shown in US dollars.
  - Confirm that when unfocusing the input field, the number is automatically formatted with the selected currency and decimals (if "Show decimals" is off, then the number is rounded to the nearest integer, if on, then it's shown to the nearest hundredth). Confirm that the same is true on the front-end.
  - Choose various currencies and test the symbols and formatting in both the editor and front-end.
4. Change your browser language to something that will result in different number formatting than your usual language. For example, my Firefox is usually set to `en-US`, so I set it to `fr-FR` and restarted the browser app.
5. Edit the Marketplace listing with your Price block and confirm that the number shown is formatted according to your browser's language setting. For example, before setting my Firefox to `fr-FR`, the number 500000 in EUR is shown as `€500,000.00`. After changing to French, it's shown as `500 000,00 €`.
6. Update the post and view on the front-end, and confirm the same. Verify that the format shown on the front-end matches the formatting last saved, _not_ the reader's browser language.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/class-newspack-listings-taxonomies.php
Comment (adekbadek) [inline includes/class-newspack-listings-taxonomies.php:17]: Not really ;)

File: includes/class-newspack-listings-taxonomies.php
Comment (adekbadek) [inline includes/class-newspack-listings-taxonomies.php:179]: Why is only one of the four listing types handled here? After reverting a generic listing to draft, it was still available in the taxonomy sidebar on a post.

**Review by adekbadek — CHANGES_REQUESTED** [review]
The core functionality, as well as the new block, work well. There are some issues though - apart from the ones in separate comment:

1. The list of the assignable listings should not contain the edited listing. 
2. In the taxonomy panel in the sidebar, after clicking "Add new <listing-type>", the text above the input says: "New Category Name". I'd expect it to say "New <listing-type>". After filling out this new listing name, confirming, and publishing, the listing is not created. I think the "Add new…" should rather link to editor creating a new listing.

File: includes/class-newspack-listings-taxonomies.php
Comment (dkoo) [reply includes/class-newspack-listings-taxonomies.php:417]: The `.sponsor-bio` styles are actually set in the Newspack theme, and it makes the image a tiny bit larger than the regular `.author-bio` styles. It also removes the border radius to make the avatars appear circular (which shouldn't apply to these listings). I used these existing theme styles so that these elements would hopefully inherit styles from the child themes as well as theme customizations such as colors, but I'm open to rendering markup and styles specific to the plugin if you think that would be a better solution.

File: includes/class-newspack-listings-taxonomies.php
Comment (dkoo) [reply includes/class-newspack-listings-taxonomies.php]: Whoops, fixed in f2d554c

File: src/blocks/price/edit.js
Comment (dkoo) [reply src/blocks/price/edit.js]: Fixed in 621c497

File: src/blocks/price/view.php
Comment (dkoo) [reply src/blocks/price/view.php]: Whoops, missed this one because I was looking for "event". 😓  Fixed in a70feef

File: includes/class-newspack-listings-taxonomies.php
Comment (dkoo) [reply includes/class-newspack-listings-taxonomies.php]: Fixed in a70feef

File: src/blocks/price/edit.js
Comment (dkoo) [reply src/blocks/price/edit.js]: Nice catch—fixed in 9ae4141

File: includes/class-newspack-listings-block-patterns.php
Comment (dkoo) [reply includes/class-newspack-listings-block-patterns.php:152]: Strange, this was working for me last week but no longer. I find that registering the patterns on `init` (instead of `admin_init`) seem to fix it for me, what about you? d27b19a

File: includes/class-newspack-listings-taxonomies.php
Comment (dkoo) [reply includes/class-newspack-listings-taxonomies.php:197]: This was an oversight (since I started off with just one post type for testing). Should be fixed in a6f0195.

**Comment by dkoo** [comment]
> The list of the assignable listings should not contain the edited listing.

I couldn't figure out a more elegant solution than a50068a, do you have any better suggestions?

> In the taxonomy panel in the sidebar, after clicking "Add new ", the text above the input says: "New Category Name". I'd expect it to say "New ". After filling out this new listing name, confirming, and publishing, the listing is not created. I think the "Add new…" should rather link to editor creating a new listing.

Good idea—updated in ce6aa70.

File: includes/class-newspack-listings-block-patterns.php
Comment (adekbadek) [reply includes/class-newspack-listings-block-patterns.php:152]: It does!

File: includes/class-newspack-listings-taxonomies.php
Comment (adekbadek) [inline includes/class-newspack-listings-taxonomies.php:325]: Looks like `$new_term` is an array, not object:

```
PHP Notice:  Trying to get property 'term_id' of non-object
```

File: includes/class-newspack-listings-taxonomies.php
Comment (adekbadek) [reply includes/class-newspack-listings-taxonomies.php:200]: Not sure if that worked, I've found this issue:
1. Create a Place
2. Assign it to a Marketplace
3. Rename the place
4. Edit the Marketplace - there is now an additional Place taxonomy item - both old and new name are listed are assignable taxonomy options

Comment (adekbadek) [reply includes/class-newspack-listings-taxonomies.php:200]: Also, the comment still mentions "Place" ;)

File: src/editor/shadow-taxonomies/index.js
Comment (adekbadek) [inline src/editor/shadow-taxonomies/index.js:29]: Enhancement: it might take a moment for the taxonomies to load - maybe a spinner could be shown in the meantime?

**Review by adekbadek — CHANGES_REQUESTED** [review]
Sorry, something made GH submit the review 👆 before I wanted it to 🤷‍♀️

> The .sponsor-bio styles are actually set in the Newspack theme, 

It looks ok now, for some reason (perhaps theme was out of date locally) the image was extremely small before.

> I couldn't figure out a more elegant solution than a50068a, do you have any better suggestions?

Hm, this looks pretty hacky. Maybe there's a way to do it in the editor? The listing linking could even be a separate custom panel, which would make the sidebar more readable.

**Comment by dkoo** [comment]
> Not sure if that worked, I've found this issue:
> 1. Create a Place
> 2. Assign it to a Marketplace
> 3. Rename the place
> 4. Edit the Marketplace - there is now an additional Place taxonomy item - both old and new name are listed are assignable taxonomy options

Ah, I see the problem. We were looking up shadow terms by post title, so if you update the post title, it thinks a term doesn't exist, so we go and create it.

In 4f374be I added a second check to look for shadow terms by slug. So now the term is only created if we can't find a shadow term by both post title and post slug.

As part of this testing I also saw a lot of "orphaned" terms getting created—terms that aren't connected to a post anymore. In the same commit I added an admin handler to look for any terms that don't have a corresponding post and delete them.

**Comment by dkoo** [comment]
> Maybe there's a way to do it in the editor? The listing linking could even be a separate custom panel, which would make the sidebar more readable.

Are you thinking a single custom sidebar panel to hold all of these listing types? Maybe an autocomplete field for each type that can be associated with the current post?

**Comment by adekbadek** [comment]
Yes, that's way better 🙌

**Comment by dkoo** [comment]
Thanks for the feedback and suggestions. I'm going to put this PR on hold for a bit while I spend some time building out this custom UI.

**Comment by dkoo** [comment]
@adekbadek After some thought I realized that listings don't just have a single type of relationship with each other and other posts/pages—it's definitely more of a parent/child relationship. I also struggled to think of a real use-case for applying CPTs to others of the same type (e.g. a Marketplace Listing owning another Marketplace Listing), so this is now no longer part of the UI. I still use the shadow taxonomy to apply only the shadow term for each listing CPT to itself, so we can query those and create missing terms and delete orphaned terms, but now this is the only purpose for being able to apply shadow terms to their own post types.

I split the UI into two panels to reflect these changes:

Screen Shot 2021-04-16 at 4 19 35 PM

We don't currently show the child listings on the parent's front-end, but I would like to add a grid-based component to show these. For example, if you have a Place representing a real estate firm which is applied as a parent to several Marketplace Listings that represent its property listings, you should optionally see a grid or list of those property listings somewhere on the real estate firm's page. We can maybe tackle this in a separate PR since this one is getting big.

Some other notes:

* There are two new global options (and per-post overrides) to hide parent and child listings that let admins choose whether these are displayed on the front-end. Both options are disabled by default, meaning parents and children will be shown on the front-end (once implemented).

* I changed "Add new (post type)" to a "View All" link instead, because I felt this is more useful if you have a lot of listings and can't remember the exact title to search for.

* When editing a listing CPT, the parent/child autocomplete fields are shown for only the post types that apply to the currently edited listing. So for Places, for example, they can own listings of any type, pages, and posts, but no other listing can own Places, so you'll only see the "Child Listings" panel for those listing types. Generic Listings can't own any posts other than regular posts or pages, so you'll only see those post types in its Child Listings panel. Any listing type can be applied to a regular post or page, but since posts and pages can't be applied as a parent to listings, that's a simpler relationship and the UI accordingly calls this "Related Listings" when editing a regular post or page.

* Another challenge with the UI I ran into was that when applying parent listings to a post, these are treated as taxonomy terms in the editor, so they're not actually saved until you "Save" or "Update" the post. However, applying child posts uses a custom REST endpoint that looks up each post in the autocomplete field and applies the current post's shadow term to all of them (or removes it from those that are removed from the UI). Because this happens separately from the regular post editor functionality, these updates happen immediately. I added some messaging to notify the user when the update happens, but it does feel somewhat strange to have two similar-looking UIs behave differently. We could potentially store child post IDs in a meta field and then update each child post on `insert_post`, but I worry that would overcomplicate an already pretty complicated thing, and I worry about the post meta getting out of sync with which posts actually have the terms (e.g. when terms get deleted). Another potential solution would be to use a custom REST endpoint to assign parents as well, but that seems overkill since the standard editor functionality works just fine for that.

I welcome your feedback, suggestions, and testing for this UI. Also paging @thomasguillot since this is UI-heavy work, in case you have any design-related feedback or want to take a crack at designing something different.

This is clearly a work-in-progress, so I'm adding a WIP label to this PR to reflect that.

File: includes/class-newspack-listings-api.php
Comment (adekbadek) [inline includes/class-newspack-listings-api.php:96]: This should be updated - it's a POST request.

File: src/editor/shadow-taxonomies/index.js
Comment (adekbadek) [inline src/editor/shadow-taxonomies/index.js:173]: When editing a marketplace, I see posts & pages field. After inputing text, nothing is found and the request returns `404`. 

image

**Review by adekbadek — CHANGES_REQUESTED** [review]
> When editing a listing CPT, the parent/child autocomplete fields are shown for only the post types that apply to the currently edited listing. So for Places, for example, they can own listings of any type, pages, and posts, but no other listing can own Places […]

This is getting complex - adding some tests would come a long way in demystifying those relations.

**Comment by dkoo** [comment]
> This is getting complex - adding some tests would come a long way in demystifying those relations.

It is getting complex, isn't it? I agree that this plugin is in need of some test coverage, but I feel like adding them here would make this PR even bigger and harder to review. Should we consider adding a test suite in a separate PR after this lands? Since we're not release Phase 2 features until all are complete, we'd have a chance to catch any errors with this PR again once the tests are implemented.

File: src/editor/index.js
Comment (dkoo) [reply src/editor/index.js:36]: I'm not able to replicate this. Just in case, could you `npm ci` or `npm install` and try again after rebuilding?

File: src/editor/index.js
Comment (dkoo) [reply src/editor/index.js]: Nope, removed in df7843f

File: includes/class-newspack-listings-api.php
Comment (dkoo) [reply includes/class-newspack-listings-api.php]: Ah yes, updated in a3978bd

File: src/editor/shadow-taxonomies/index.js
Comment (dkoo) [reply src/editor/shadow-taxonomies/index.js:184]: Fixed in 527fe4d—forgot that the core `posts` and `pages` endpoint have the plural `s`.

Also in the same commit: since only published listings get shadow terms, we now hide the Child Listings UI if the post you're editing isn't published. Also added a `validate_post` method to the taxonomies class to make it a bit easier to check whether a post should have a shadow term.

File: src/editor/index.js
Comment (adekbadek) [reply src/editor/index.js:36]: Whoops, can't replicate anymore. I was thinking there might be an issue with block registration, but it must've been something else.

File: src/editor/shadow-taxonomies/index.js
Comment (adekbadek) [reply src/editor/shadow-taxonomies/index.js:184]: The request now works, but I see an error in the logs: 
```
Trying to get property 'term_id' of non-object
```
Originating from here: 

https://github.com/Automattic/newspack-listings/blob/527fe4d35e4b1df818a4c41a8e10d656c8afa8bf/includes/class-newspack-listings-taxonomies.php#L578

Where `$shadow_term` is false, when I'm linking a just-published Event to a post or a page.

**Review by adekbadek — CHANGES_REQUESTED** [review]
(commented in existing thread)

**Comment by adekbadek** [comment]
> Should we consider adding a test suite in a separate PR after this lands? Since we're not releasing Phase 2 features until all are complete, we'd have a chance to catch any errors with this PR again once the tests are implemented.

Since this is not merged to `master`, the tests can be added later, but I'd be wary of merging Phase 2 as a single huge PR. This would introduce a lot of potential of errors.

**Comment by dkoo** [comment]
>  I'd be wary of merging Phase 2 as a single huge PR. This would introduce a lot of potential of errors.

Definitely something I've thought about, but the features outlined for a Phase 2 release don't make much sense to roll out individually. The approach we've taken in reviewing and merging smaller PRs against a clone of `master` (while minimizing updates to `master` in the meantime) should hopefully minimize the errors when we finally merge that branch.

File: src/editor/shadow-taxonomies/index.js
Comment (dkoo) [reply src/editor/shadow-taxonomies/index.js:184]: I made some logical errors in the `validate_post` method in the prior commit. Should be fixed in 4b43a7c.

File: includes/class-newspack-listings-taxonomies.php
Comment (adekbadek) [inline includes/class-newspack-listings-taxonomies.php:240]:
```suggestion
		// Post must not be published.
```
Since the check below was updated. But that's weird, because I'd expect a published post to be linkable to a listing. Is that check below intended?

Comment (dkoo) [reply includes/class-newspack-listings-taxonomies.php:240]: It's somewhat confusing (I even confused myself!) because `validate_post` returns `true` if the listing is "valid" (e.g. it should have a shadow term), and `false` if it's not valid. So the check should be: if the post is **not** published, return `false`, which is what the change in 4b43a7c reflects.

Comment (adekbadek) [reply includes/class-newspack-listings-taxonomies.php:240]: Then the word "validate" is perhaps not the best - how about `should_update_shadowterm` as the method name?
Also, the code comment should still be updated ;)

Comment (dkoo) [reply includes/class-newspack-listings-taxonomies.php]: `should_update_shadow_term` does make the function's purpose clearer. Updated in a136170.

**Review by adekbadek — APPROVED** [review]

**Comment by dkoo** [comment]
@thomasguillot Since your redesigned UI won't affect the back-end portion of this PR, and this isn't getting released until all phase 2 PRs are approved, I'm going to merge this and handle the new UI in a separate PR.

**Pull request merged**
